Automattic/newspack-popups
# feat: refactor segments into a taxonomy

### All Submissions:






### Changes proposed in this Pull Request:

This PR is the part 1 of the Segments refactor, in which we change how we store segments. Instead of a single array in an option, we now use a taxonomy.

It relies heavily on the tests introduced in https://github.com/Automattic/newspack-popups/pull/1085 to make sure we maintain the same interface as before and the rest of our code can still work without further changes.

There's only one behavior I had to change - because it was hard and pointless to maintain the former buggy behavior.

Before this PR, the "priority" attribute of segments were not updated when you created a segment. In fact, it was ignored and what really mattered was the position of the segment in the array. When you called `get_segments` and we detected one segment missing that property, we would reindex and re-populate the priority for all segments, but that change was not persisted.

When you created another segment, and there was a previous segment without the priority property, all the segments would kind of be accidentally reindexed, because `create_segment` would call `get_segments`, that would reindex the segments, and when we updated the option with the new segment, the new priorities would be persisted.

Now, the behavior changed a bit.

* The informed priority on segment creation is still ignored (same behavior). The segment is always added to the end of the list (lowest priority)
* Every time a segment is created, we add the `priority` property, so priorities are always informed and indexed starting from zero (Before, if you informed a random priority, say `10`, to a segment on creation, this priority value would still be persisted in the segment.

The changes in the tests reflect this small change.

Next step on the refactor will be:
* Associate prompts to segments using the taxonomy and not post_meta
* validate segments before saving them
* Implement a better interface with the rest of the code (instead of an array, an object of a `Segment` class representing each segment)
* optimize and simplify the sorting methods

### How to test the changes in this Pull Request:

1. Make sure this plugin is running on `master` branch
2. Set up the site, or using an existing site, with many segments and prompts
3. Checkout this branch
4. Browse the admin and make sure all segments and prompts look the same
5. Make sure prompts are correctly segmented on the front-end
6. Test adding and editing segments
7. Test assigning prompts to segments
8. Test everything you can think about segments

### Other information:
